Prefer retain over CFRetain and copy for ObjC Callbacks

LumioseSil opened the pull request:
#### 0b25a1815a194332b57e94a2a0473efcfb632799
<pre>
Prefer retain over CFRetain and copy for ObjC Callbacks
<a href="https://bugs.webkit.org/show_bug.cgi?id=252575i">https://bugs.webkit.org/show_bug.cgi?id=252575i</a>

Given the nature of this file, we have to manually
manage memory anyway.

* Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
  (objCCallbackFunctionForMethod): Send retain message
  to class object instead of retaining it in a hacky fashion.
  (objCCallbackFunctionForBlock): Retain block instead
  of copying it.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b25a1815a194332b57e94a2a0473efcfb632799

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1513 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2432 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1668 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1602 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1443 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1526 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1378 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2269 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1381 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1362 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1301 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1292 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1377 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1401 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2462 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1455 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1423 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1269 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1552 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1339 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1368 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/334 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1475 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1594 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/447 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
```diff
@@ -727,13 +727,13 @@ JSObjectRef objCCallbackFunctionForMethod(JSContext *context, Class cls, Protoco
-        [invocation setTarget:cls];
-        // We need to retain the target Class because m_invocation doesn't retain it by default (and we don't want it to).
-        // FIXME: What releases it?
-        CFRetain((__bridge CFTypeRef)cls);
```
> ensuring memory is not leaked, as nothing that releases the Class object could be found

Your patch does not change anything about the lifetime here; both retain and copy result in +1 retain count. And, I don't think preferring copy over retain makes any sense, nor does deleting the comment ("what releases it?") that is clearly still relevant.

AZero13 replied:
Wouldn't ARC take care of that though with copy?

AZero13 replied:
void destroy(Heap& heap) seems to release the memory too. Hopefully I am not wrong on this

LumioseSil committed:
Prefer retain over CFRetain and copy for ObjC Callbacks
https://bugs.webkit.org/show_bug.cgi?id=252575i

Given the nature of this file, we have to manually
manage memory anyway.

* Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
  (objCCallbackFunctionForMethod): Send retain message
  to class object instead of retaining it in a hacky fashion.
  (objCCallbackFunctionForBlock): Retain block instead
  of copying it.
